### ISSUE TITLE
Binding at startup; can't unbind correctly

### ISSUE THREAD

**Issue by mysteryx93** [bryansteiner/gpu-passthrough-tutorial#31]
Using Garuda Linux on laptop with NVidia dual-graphics. It seems there's something in Garuda's configuration that hangs onto the GPU and hangs while detaching.

The vfio-pci approach also wasn't working.

The solution is to add these to `/etc/mkinitcpio.conf` before the list of video drivers, to load vfio-pci early.

    MODULES=(crc32c-intel intel_agp i915 vfio_pci vfio vfio_iommu_type1 vfio_virqfd amdgpu radeon nouveau)

then generate configs with

    mkinitcpio -P

Now I boot Linux with an unbound GPU, great!

From here, your bind and unbind scripts run successfully, but I remain unable to use the GPU in Linux.

I run `unbind_vfio.sh` and then

```
> nvidia-xconfig --query-gpu-info

Number of GPUs: 1

GPU #0:
Name      : NVIDIA GeForce RTX 2060
UUID      : GPU-5b17f5ee-5f05-94a6-949f-d6a2c92e22d4
PCI BusID : PCI:1:0:0

Number of Display Devices: 1

Display Device 0 (TV-2):
EDID Name             : SAMSUNG
Minimum HorizSync     : 15.000 kHz
Maximum HorizSync     : 81.000 kHz
Minimum VertRefresh   : 24 Hz
Maximum VertRefresh   : 75 Hz
Maximum PixelClock    : 230.000 MHz
Maximum Width         : 1920 pixels
Maximum Height        : 1080 pixels
Preferred Width       : 1920 pixels
Preferred Height      : 1080 pixels
Preferred VertRefresh : 60 Hz
Physical Width        : 1020 mm
Physical Height       : 570 mm
```

Linux detects the GPU. It even detects the TV plugged via HDMI. For some reason I'm unable to use it. Plugging the HDMI cable, the "switch display" bar appears, but I'm able to switch output. The TV doesn't appear in Display Configuration, and I'm unable to start Optimus Manager.

Any idea how to get the GPU in a functional state from here?

By editing `/etc/mkinitcpio.conf` as above, everything works normal.

Editing `/etc/default/grub`

    GRUB_CMDLINE_LINUX_DEFAULT="quiet splash rd.udev.log_priority=3 vt.global_cursor_default=0 systemd.unified_cgroup_hierarchy=1 loglevel=3 intel_iommu=on vfio-pci.ids=10de:1f15,10de:10f9"

Rebuilding config

    sudo grub-mkconfig -o /boot/grub/grub.cfg

Rebooting.

HDMI output doesn't work (good!).

Running `unbind_vfio.sh`. Success.

I remain with the behavior described above. GPU doesn't work.

Let's try the VM. From a bound state, the VM runs without error. It only displays "Connecting to graphical console for guest" and fails to take the HDMI output.

Removed the display device, now I get a console with

```
BdsDxe: loading Boot0003 "Windows Boot Manager" from HD(1,GPT,D0C359A4-6867-44B3-B3E7-D2F1F3644DE7,0x800,0x32000)/\EFI\Microsoft\Boot\bootmgfw.efi
BdsDxe: starting Boot0003 "Windows Boot Manager" from HD(1,GPT,D0C359A4-6867-44B3-B3E7-D2F1F3644DE7,0x800,0x32000)/\EFI\Microsoft\Boot\bootmgfw.efi
```

**Comment by mysteryx93** [comment]
I found out that simply changing the load order in `/etc/mkinitcpio.conf`, without adding any kernel parameter, causes the same problem: Linux detects the GPU and TV, but can't use HDMI output nor load Optimus Manager.

Thus the problem isn't related to your script. There's something else going on.

**Comment by mysteryx93** [comment]
After 5 days of digging into this.

I figured that I'm using an Optimus graphic card (GeoForce 2060 GTX) on a MUXless laptop. Very hard to setup.

Search "Optimus GPU passthrough" and you'll get relevant info.

But the only guide is this
https://gist.github.com/Misairu-G/616f7b2756c488148b7309addc940b28

It requires installing Bumblebee, which will result in a performance loss in Linux, and according to a friend, HDMI won't even work if I manage to get it setup.

I'm working on an alternative solution. Script that

1. optimus-manager --switch integrated
2. modprobe -r nvidia
3. virsh nodedev-detach
4. modprobe vfio

I did manage to get binding to work, but only from integrated mode.

and unbinds doing the reverse. Will kill the session but should work. Still working on that script.
